nixos/modules/config/system-path.nix

```diff
@@ -6,6 +6,34 @@
 with lib;
 
 let
+  pkgsToAttrs =
+    let
+      mkEntry =
+        let
+          isDefaultOutput = drv: drv.outputName == lib.head drv.outputs;
+          mkName = drv: if isDefaultOutput drv then drv.name else "${drv.name}:${drv.outputName}";
+        in
+        drv: lib.nameValuePair (mkName drv) {
+          enable = true;
+          package = drv;
+        };
+    in
+    ps: lib.listToAttrs (map mkEntry ps);
+
+  packagesByPrio = types.coercedTo
+    (types.listOf types.packageByPrio)
+    pkgsToAttrs
+    (types.attrsOf (types.submodule ( { lib, ... }: {
```

Comment:
If used for `environment.systemPackages`, this is not entirely backwards compatible, because the result of this option is now an attribute set instead of a list, so any module making use of this option will break.

For `environment.systemPackages` the coercion should go the other way instead, which then also removes the need for the `pkgsToAttrs` function.

Reply:
But if we coerce in the other direction, then we lose all the benefit, since then we still end up with a list and we are in the same situation as today.

In the whole of nixpkgs, there was only one place where the value of `environment.systemPackages` is used, but it is being set from everywhere. And I also expect to see the same in peoples configs, that people just set packages but don't use the resulting value themselves.

So I think what we want is that the actual value of the option is always an attrset, so that we can have the overrides and such, but maintain the possibility of setting the option using a list, to maintain backwards compat, which is what the current type does.

Reply:
Yeah, I still doubt that anyone directly consumes this option since `system.path` is almost always what you'd want instead, but I guess it wouldn't hurt to add `apply`.

I'd like to avoid introducing a new option since it makes things even more complex, and might require people to understand a lot of internals to understand how to override a package on the system path.

Reply:
I've added an apply, this seems to work nicely, thanks for the suggestion!